### ISSUE TITLE
Restricting Field arithmetic to smaller subgroups

### ISSUE THREAD

**Issue by geostergiop** [mhostetter/galois#273]
Hi Matt, I have a somewhat non-conventional question. Is there a way to restrict Field arithmetic to smaller subgroups than the ones defined when constructing a GF(2^m) field? E.g a field `GF = galois.GF((2**233), irreducible_poly=poly, primitive_element=2, verify=False)` will reduce all arithmetic modulo the irreducible poly, but is there a way to set another element of the field and perform arithmetic modulo that element inside GF? e.g. `x * y == np.multiply(x, y)` but get the result `mod e`, where `e` an element represented as a polynomial in GF.

Let me know if this is not clear, I know it's a weird question. Has to do with an implementation of ours.
Thanks!

**Comment by mhostetter** [comment]
Hey, George. I'm not completely sure what you're trying to do...

Are you really seeking `(x * y) % e` where `x`, `y`, and `e` are all elements of `GF(2^m)`? If so, the answer is always zero since every non-zero element of a finite field perfectly divides every other non-zero element.

Or are you trying to do polynomial multiplication of `x` and `y` and polynomial remainder division with `e`? If so, could you use `galois.Poly` objects? Here's an example in `GF(2^8)`.

```python
In [1]: import galois

In [2]: GF = galois.GF(2**8, display="poly")

# An element x of GF(2^8)
In [3]: x = GF.Random(); x
Out[3]: GF(α^7 + α^6 + α^3 + α^2 + 1, order=2^8)

# An element y of GF(2^8)
In [4]: y = GF.Random(); y
Out[4]: GF(α^3 + α^2 + α + 1, order=2^8)

# x * y in GF(2^8)
In [5]: x * y
Out[5]: GF(α^6 + α^5 + α^4 + α^3 + α^2 + α + 1, order=2^8)

# An element e of GF(2^8)
In [6]: e = GF.Random(); e
Out[6]: GF(α^7 + α^6 + α^4 + α + 1, order=2^8)

# (x * y) % e in GF(2^8), always zero
In [7]: (x * y) % e
Out[7]: GF(0, order=2^8)

# x * y as polynomials over GF(2), NOTE different from x * y in GF(2^8)!!
In [8]: galois.Poly(x.vector()) * galois.Poly(y.vector())
Out[8]: Poly(x^10 + x^3 + x + 1, GF(2))

# (x * y) % e as polynomials over GF(2)
In [9]: (galois.Poly(x.vector()) * galois.Poly(y.vector())) % galois.Poly(e.vector())
Out[9]: Poly(x^6 + x^5 + x^4 + x^3 + 1, GF(2))
```

You can convert the final answer back to a `GF(2^8)` element, if you'd like.

```python
In [14]: a = (galois.Poly(x.vector()) * galois.Poly(y.vector())) % galois.Poly(e.vector()); a
Out[14]: Poly(x^6 + x^5 + x^4 + x^3 + 1, GF(2))

In [15]: GF.Vector(a.coefficients(8))
Out[15]: GF(α^6 + α^5 + α^4 + α^3 + 1, order=2^8)
```

Is that remotely similar to the calculation you're trying to perform?

**Comment by geostergiop** [comment]
Thanks, I was looking for 
```
# (x * y) % e as polynomials over GF(2)

In [9]: (galois.Poly(x.vector()) * galois.Poly(y.vector())) % galois.Poly(e.vector())
Out[9]: Poly(x^6 + x^5 + x^4 + x^3 + 1, GF(2))

In [14]: a = (galois.Poly(x.vector()) * galois.Poly(y.vector())) % galois.Poly(e.vector()); a
Out[14]: Poly(x^6 + x^5 + x^4 + x^3 + 1, GF(2))
```
in an effort to analyse GF(2^m), m=128 or 256 as a composite field of GF(2^m)^m...)^m) (no need to baffle you with these things, if you are interested there is a small [paragraph here explaining so](https://en.wikipedia.org/wiki/Finite_field_arithmetic#Composite_field))

TLDR; We are having some trouble with the output of `.minimal_poly()` and the irreducible polynomial used to construct the underlying GF(2^m). Changing the irr.,primitive polynomial used to construct GF completely changes outputs in `.minimal_poly()` for all elements, which is expected I guess although we cannot understand the algorithmic concept why this is so. Makes a mess in our implementations.

Just writing these down in case you are interested, since they are all related to your work here.
Best!

**Comment by mhostetter** [comment]
I don't know if this helps at all, but...

The minimal polynomial of `a` in `L / K` is a polynomial over `K` where `f(a) = 0` evaluated in `L`. Since the polynomial evaluation is in `L`, the arithmetic is dependent on the irreducible polynomial of `L`.

Also, the minimal polynomial of the primitive element is the (primitive) irreducible polynomial.

Here are two examples in `GF(2^8)` with different irreducible polynomials.

```python
In [1]: import galois

In [2]: poly1 = galois.primitive_polys(2, 8)[0]; poly1
Out[2]: Poly(x^8 + x^4 + x^3 + x^2 + 1, GF(2))

In [3]: GF = galois.GF(2**8, irreducible_poly=poly1, display="poly")

In [4]: print(GF.properties)
GF(2^8):
  characteristic: 2
  degree: 8
  order: 256
  irreducible_poly: x^8 + x^4 + x^3 + x^2 + 1
  is_primitive_poly: True
  primitive_element: x

In [5]: a = GF.primitive_element; a
Out[5]: GF(α, order=2^8)

In [6]: f = a.minimal_poly(); f
Out[6]: Poly(x^8 + x^4 + x^3 + x^2 + 1, GF(2))

In [7]: f(a, field=GF)
Out[7]: GF(0, order=2^8)
```

```python
In [1]: import galois

In [2]: poly2 = galois.primitive_polys(2, 8)[1]; poly2
Out[2]: Poly(x^8 + x^5 + x^3 + x + 1, GF(2))

In [3]: GF = galois.GF(2**8, irreducible_poly=poly2, display="poly")

In [4]: print(GF.properties)
GF(2^8):
  characteristic: 2
  degree: 8
  order: 256
  irreducible_poly: x^8 + x^5 + x^3 + x + 1
  is_primitive_poly: True
  primitive_element: x

In [5]: a = GF.primitive_element; a
Out[5]: GF(α, order=2^8)

In [6]: f = a.minimal_poly(); f
Out[6]: Poly(x^8 + x^5 + x^3 + x + 1, GF(2))

In [7]: f(a, field=GF)
Out[7]: GF(0, order=2^8)
```

**Comment by geostergiop** [comment]
Thanks Matt, indeed I am aware of these. 
The problem lies somewhere in the arithmetic that is dependent on the irreducible polynomial of L but I am afraid it goes well beyond engineering and straight into theoretical mathematics about finite groups in fields. 

Anyway, best!
